src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/Unsigned5.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+package sun.jvm.hotspot.utilities;
+
+import java.io.PrintStream;
+
+import sun.jvm.hotspot.debugger.*;
+import sun.jvm.hotspot.runtime.*;
+import sun.jvm.hotspot.types.*;
+
+/**
+ * Decompression algorithm from utilities/unsigned5.hpp.
+ */
+public class Unsigned5 {
+  static {
+    VM.registerVMInitializedObserver(new Observer() {
+        public void update(Observable o, Object data) {
+          initialize(VM.getVM().getTypeDataBase());
+        }
+      });
+  }
+
+  private static void initialize(TypeDataBase db) {
+    if (Assert.ASSERTS_ENABLED) {
+      Assert.that(!VM.getVM().isCore(), "Debug info not used in core build");
+    }
+
+    Integer bval = db.lookupIntConstant("UNSIGNED5::MAX_LENGTH", false);
+    int hval, lval, xval;
+    if (bval == null) {
+      // older vmStructs does not record this information; it's hardwired
+      System.out.println("[Unsigned5: defaulting BHLX values...]");//@@
+      bval = 5; hval = 64; lval = 192; xval = 0;
+    } else {
+      hval = db.lookupIntConstant("UNSIGNED5::H").intValue();
+      lval = db.lookupIntConstant("UNSIGNED5::L").intValue();
+      xval = db.lookupIntConstant("UNSIGNED5::X").intValue();
```

Comment:
These constants aren't likely to change ever, are they?  I don't see why we'd need the SA to get these constants from the VM.  It could just hard-code them like you have here.  Then UNSIGNED5 doesn't have to make VMStructs a friend class.
Also, there isn't a compatibility issue with older SA - it might work sometimes but the SA version should match the JVM that is being debugged.